File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/Root.js
// FILE: resources/kalitech/engine/bootstrap/Root.js
"use strict";

const ROOT_KEY = "__kalitech";

function getRoot() {
    if (!globalThis[ROOT_KEY]) globalThis[ROOT_KEY] = Object.create(null);
    return globalThis[ROOT_KEY];
}

function ensureRootState(K) {
    if (!K.modules) K.modules = Object.create(null);
    if (!K.instances) K.instances = Object.create(null);
    if (!K.meta) K.meta = Object.create(null);
    if (!K.instancesMeta) K.instancesMeta = Object.create(null);
    if (!K.moduleIds) K.moduleIds = Object.create(null);

    if (!K._engine) K._engine = null;
    if (!K._engineAttached) K._engineAttached = false;

    if (!K._deferred) K._deferred = [];
    if (!K._once) K._once = Object.create(null);

    if (!K.config) K.config = Object.create(null);

    if (!K.dataConfig) K.dataConfig = Object.create(null);
    if (!K.dataConfigApi) K.dataConfigApi = null;

    // NEW: controllers service
    if (!K.controllers) K.controllers = Object.create(null); // name -> { name, defs }
    if (!K.controllersApi) K.controllersApi = null;
    if (!K.controllersRegs) K.controllersRegs = [];

    return K;
}

module.exports = {getRoot, ensureRootState};

File: kalitech-engine/src/main/resources/kalitech/engine/bootstrap/Root.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Classes = luaRuntime.class
ROOT_KEY = "__kalitech"
function getRoot(self)
    if not _G[ROOT_KEY] then
        _G[ROOT_KEY] = KObject:create(nil)
    end
    return _G[ROOT_KEY]
end
function ensureRootState(self, K)
    if not K.modules then
        K.modules = KObject:create(nil)
    end
    if not K.instances then
        K.instances = KObject:create(nil)
    end
    if not K.meta then
        K.meta = KObject:create(nil)
    end
    if not K.instancesMeta then
        K.instancesMeta = KObject:create(nil)
    end
    if not K.moduleIds then
        K.moduleIds = KObject:create(nil)
    end
    if not K._engine then
        K._engine = nil
    end
    if not K._engineAttached then
        K._engineAttached = false
    end
    if not K._deferred then
        K._deferred = {}
    end
    if not K._once then
        K._once = KObject:create(nil)
    end
    if not K.config then
        K.config = KObject:create(nil)
    end
    if not K.dataConfig then
        K.dataConfig = KObject:create(nil)
    end
    if not K.dataConfigApi then
        K.dataConfigApi = nil
    end
    if not K.controllers then
        K.controllers = KObject:create(nil)
    end
    if not K.controllersApi then
        K.controllersApi = nil
    end
    if not K.controllersRegs then
        K.controllersRegs = {}
    end
    return K
end
local BootstrapRootApi = Classes:create()
BootstrapRootApi.name = "BootstrapRootApi"
BootstrapRootApi.prototype.getRoot = getRoot
BootstrapRootApi.prototype.ensureRootState = ensureRootState
return Classes:construct(BootstrapRootApi)
